File: Backend/ShawarmaBackend/Services/Implementations/MenuService.cs
using ShawarmaBackend.Models.Domain;
using ShawarmaBackend.Models.DTOs.Requests;
using ShawarmaBackend.Repositories.Interfaces;
using ShawarmaBackend.Services.Interfaces;

namespace ShawarmaBackend.Services.Implementations
{
    public class MenuService : IMenuService
    {
        private readonly IMenuRepository _menuRepository;

        public MenuService(IMenuRepository menuRepository)
        {
            _menuRepository = menuRepository;
        }

        public async Task<List<MenuItem>> GetAllMenuItems()
        {
            return await _menuRepository.GetAllMenuItems();
        }

        public async Task<List<MenuItem>> GetAvailableMenuItems()
        {
            return await _menuRepository.GetAvailableMenuItems();
        }

        public async Task<MenuItem> GetMenuItemById(int itemId)
        {
            return await _menuRepository.GetMenuItemById(itemId);
        }

        public async Task<(int itemId, string message)> AddMenuItem(AddMenuItemRequest request)
        {
            // Validate request
            if (request.Price <= 0)
            {
                return (-1, "Price must be greater than 0");
            }

            if (request.Quantity < 0)
            {
                return (-1, "Quantity cannot be negative");
            }

            var menuItem = new MenuItem
            {
                Name = request.Name,
                Category = request.Category,
                Price = request.Price,
                Quantity = request.Quantity,
                ImagePath = request.ImagePath,
                IsAvailable = request.IsAvailable
            };

            return await _menuRepository.AddMenuItem(menuItem);
        }

        public async Task<(int itemId, string message)> UpdateMenuItem(UpdateMenuItemRequest request)
        {
            // Validate item exists
            var existingItem = await _menuRepository.GetMenuItemById(request.ItemID);
            if (existingItem == null)
            {
                return (-1, "Menu item not found");
            }

            // Validate request
            if (request.Price <= 0)
            {
                return (-1, "Price must be greater than 0");
            }

            if (request.Quantity < 0)
            {
                return (-1, "Quantity cannot be negative");
            }

            var menuItem = new MenuItem
            {
                ItemID = request.ItemID,
                Name = request.Name,
                Category = request.Category,
                Price = request.Price,
                Quantity = request.Quantity,
                ImagePath = request.ImagePath,
                IsAvailable = request.IsAvailable
            };

            return await _menuRepository.UpdateMenuItem(menuItem);
        }

        public async Task<(int itemId, string message)> DeleteMenuItem(int itemId)
        {
            // Validate item exists
            var existingItem = await _menuRepository.GetMenuItemById(itemId);
            if (existingItem == null)
            {
                return (-1, "Menu item not found");
            }

            return await _menuRepository.DeleteMenuItem(itemId);
        }

        public async Task<(int itemId, int quantity, string message)> UpdateStock(int itemId, int quantity)
        {
            // Validate item exists
            var existingItem = await _menuRepository.GetMenuItemById(itemId);
            if (existingItem == null)
            {
                return (-1, 0, "Menu item not found");
            }

            if (quantity < 0)
            {
                return (-1, 0, "Quantity cannot be negative");
            }

            return await _menuRepository.UpdateStock(itemId, quantity);
        }
    }
}

File: Frontend/Shawarma-O’Clock-frontend/shawarma-frontend/src/services/menuService.js
// src/services/menuService.js
import apiClient from './apiClient';

const menuService = {
  // Get all menu items (Admin)
  async getAllMenuItems() {
    try {
      const response = await apiClient.get('/Menu');
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to fetch menu items';
    }
  },

  // Get available menu items (Public)
  async getAvailableMenuItems() {
    try {
      const response = await apiClient.get('/Menu/available');
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to fetch available items';
    }
  },

  // Get menu item by ID
  async getMenuItemById(id) {
    try {
      const response = await apiClient.get(`/Menu/${id}`);
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to fetch item details';
    }
  },

  // Add new menu item (Admin only)
  async addMenuItem(itemData) {
    try {
      const response = await apiClient.post('/Menu', itemData);
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to add menu item';
    }
  },

  // Update menu item (Admin only)
  async updateMenuItem(itemData) {
    try {
      const response = await apiClient.put('/Menu', itemData);
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to update menu item';
    }
  },

  // Delete menu item (Admin only)
  async deleteMenuItem(id) {
    try {
      const response = await apiClient.delete(`/Menu/${id}`);
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to delete menu item';
    }
  },

  // Update stock (Admin only)
  async updateStock(id, quantity) {
    try {
      const response = await apiClient.patch(`/Menu/stock/${id}?quantity=${quantity}`);
      return response.data;
    } catch (error) {
      throw error.response?.data?.message || 'Failed to update stock';
    }
  },
};

export default menuService;